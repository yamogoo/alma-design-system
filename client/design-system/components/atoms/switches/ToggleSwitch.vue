<script setup lang="ts">
import {
  useAttrs,
  useId,
  onMounted,
  watch,
  toValue,
  computed,
  useTemplateRef,
} from "vue";
import g from "gsap";

import tokens from "@/tokens";

import { usePressed } from "@/composables/local";

import type { ToggleSwitchProps } from "@/components/atoms";

defineOptions({
  inheritAttrs: false,
});

useAttrs();

const props = withDefaults(defineProps<ToggleSwitchProps>(), {
  variant: "default",
  mode: "primary",
  size: "md",
  isActive: false,
  isDisabled: false,
  useNative: false,
});

const emit = defineEmits<{
  (e: "update:is-active", state: boolean): void;
}>();

const id = useId();

const refRoot = useTemplateRef<HTMLLabelElement | HTMLDivElement | null>(
  "root"
);
const refKnob = useTemplateRef<HTMLDivElement | null>("knob");

const { isPressed } = usePressed(refRoot);

const onChange = (e: Event): void => {
  if (props.isDisabled) return;

  const state = (e.target as HTMLInputElement).checked;
  emit("update:is-active", state);
};

const padding = computed(
  () =>
    tokens.atoms.toggleSwitch.default[`${props.size}`].elements.track.padding
      .value
);

const onKeyDown = (e: KeyboardEvent): void => {
  if (props.isDisabled) return;
  if (e.key === " " || e.key === "Enter") {
    e.preventDefault();
    emit("update:is-active", !props.isActive);
  }
};

watch(isPressed, (val) => {
  if (!val && !props.useNative) emit("update:is-active", !props.isActive);
});

/* * * Animations * * */

const onAnimateKnob = (duration = 0.25): void => {
  const width = ((refRoot.value?.clientWidth || 0) - toValue(padding) * 2) / 2;

  if (refKnob.value) {
    const knobPosX = props.isActive ? width : 0;

    g.to(refKnob.value, {
      x: knobPosX,
      duration,
      ease: "power4.out",
    });
  }
};

watch(
  () => props.isActive,
  () => {
    onAnimateKnob();
  }
);

onMounted(() => {
  if (refKnob.value) {
    onAnimateKnob(0.0);
  }
});
</script>

<template>
  <template v-if="!useNative">
    <div
      ref="root"
      :class="[
        'toggle-switch',
        `toggle-switch_variant-${variant}`,
        `toggle-switch_mode-${mode}`,
        `toggle-switch_size-${size}`,
        `toggle-switch_state-${isActive ? 'active' : 'normal'}`,
        { 'toggle-switch_disabled': isDisabled },
      ]"
      role="switch"
      :aria-labelledby="`label-${id}`"
      :aria-checked="isActive"
      :aria-disabled="isDisabled"
      tabindex="0"
      @keydown="onKeyDown"
    >
      <div class="toggle-switch__track">
        <div class="toggle-switch__track-container">
          <span ref="knob" class="toggle-switch__knob"></span>
        </div>
      </div>
      <span v-if="label" class="toggle-switch__label">{{ label }}</span>
    </div>
  </template>
  <template v-if="useNative">
    <label
      ref="root"
      :for="id"
      :class="[
        'toggle-switch',
        `toggle-switch_variant-${variant}`,
        `toggle-switch_mode-${mode}`,
        `toggle-switch_size-${size}`,
        `toggle-switch_state-${isActive ? 'active' : 'normal'}`,
        { 'toggle-switch_disabled': isDisabled },
      ]"
      :aria-labelledby="`label-${id}`"
      tabindex="0"
    >
      <input
        :id
        v-bind="$attrs"
        type="checkbox"
        :checked="isActive"
        @change="onChange"
      />
      <div class="toggle-switch__track">
        <div class="toggle-switch__track-container">
          <span ref="knob" class="toggle-switch__knob"></span>
        </div>
      </div>
      <span v-if="label" class="toggle-switch__label">{{ label }}</span>
    </label>
  </template>
</template>

<style lang="scss">
@use "sass:map";

@mixin defineSizes($map: get($atoms, "toggle-switch")) {
  @each $variant, $sizes in $map {
    @each $size, $val in $sizes {
      &_variant-#{$variant} {
        &.toggle-switch_size-#{$size} {
          $gap: get($val, "self.gap");
          $width: px2rem(get($val, "self.width"));
          $height: px2rem(get($val, "self.height"));

          $track-padding: px2rem(get($val, "elements.track.padding"));
          $label-font-style: get($val, "elements.label.font-style");

          gap: $gap;

          .toggle-switch__track {
            @include box($width, $height);
            border-radius: $height;
            padding: $track-padding;
          }

          .toggle-switch__knob {
            @include box($height);
            border-radius: calc($height / 2);
          }

          .toggle-switch__label {
            @extend %t__#{$label-font-style};
          }
        }
      }
    }
  }
}

@mixin defineThemes($map: get($themes, "light.atoms.toggle-switch")) {
  @each $mode, $modes in $map {
    &_mode-#{$mode} {
      &:not(.toggle-switch_disabled) {
        &.toggle-switch_state {
          &-normal {
            .toggle-switch__track {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.track.normal"
                );
              }
            }

            .toggle-switch__knob {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.knob.normal"
                );
              }
            }

            .toggle-switch__label {
              @include themify($themes) {
                color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.label.normal"
                );
              }
            }
          }

          &-active {
            .toggle-switch__track {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.track.active"
                );
              }
            }

            .toggle-switch__knob {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.knob.active"
                );
              }
            }

            .toggle-switch__label {
              @include themify($themes) {
                color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.label.active"
                );
              }
            }
          }
        }
      }

      &.toggle-switch_disabled {
        &.toggle-switch_state {
          &-normal {
            .toggle-switch__track {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.track.disabled"
                );
              }
            }

            .toggle-switch__knob {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.knob.disabled"
                );
              }
            }

            .toggle-switch__label {
              @include themify($themes) {
                color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.label.disabled"
                );
              }
            }
          }

          &-active {
            .toggle-switch__track {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.track.disabled"
                );
              }
            }

            .toggle-switch__knob {
              @include themify($themes) {
                background-color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.knob.disabled"
                );
              }
            }

            .toggle-switch__label {
              @include themify($themes) {
                color: themed(
                  "atoms.toggle-switch.#{$mode}.elements.label.disabled"
                );
              }
            }
          }
        }
      }
    }
  }
}

.toggle-switch {
  box-sizing: border-box;
  display: flex;
  flex-direction: row-reverse;
  align-items: center;
  cursor: pointer;
  @include box(max-content);

  @include defineSizes();
  @include defineThemes();

  input {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    white-space: nowrap;
    border-width: 0;
  }

  &__track {
    overflow: hidden;

    &-container {
      @include box(100%);
    }
  }

  &__knob {
    display: block;
  }
}
</style>
