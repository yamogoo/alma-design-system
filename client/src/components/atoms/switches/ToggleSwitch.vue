<script setup lang="ts">
import type { ToggleSwitchProps } from "./toggleSwitch";

withDefaults(defineProps<ToggleSwitchProps>(), {
  isActive: false,
  isDisabled: false,
});
</script>

<template>
  <div class="toggle-switch"></div>
</template>

<style lang="scss">
/* .toggle-switch { } */
</style>
