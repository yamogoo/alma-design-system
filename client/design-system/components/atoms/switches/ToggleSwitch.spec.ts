import { describe, expect, test } from "vitest";
import { mount } from "@vue/test-utils";

// import { getTypedEmittedEvent } from "@/utils/vitest";

import ToggleSwitch from "./ToggleSwitch.vue";

// const ACTIVE_CLASS = "toggle-switch_active",
//   NORMAL_CLASS = "toggle-switch_normal";

// const getInput = <T>(wrapper: VueWrapper<T>) => {
//   return wrapper.find('input[type="checkbox"]');
// };

describe("ToggleSwitch", () => {
  describe("props", () => {
    test.each([true, false])(
      "should render correctly with useNative=%s",
      (useNative) => {
        const wrapper = mount(ToggleSwitch, {
          props: { useNative, isActive: true },
        });

        expect(wrapper.exists()).toBe(true);

        if (useNative) {
          expect(wrapper.find('input[type="checkbox"]').exists()).toBe(true);
        } else {
          // expect(wrapper.attributes("role")).toBe("switch");
          // expect(wrapper.attributes("aria-checked")).toBe("true");
        }
      }
    );
  });

  // describe("values", () => {
  //   test("should not emit events when disabled", async () => {
  //     const wrapper = mount(ToggleSwitch, {
  //       props: { isDisabled: true, useNative: false },
  //     });

  //     await wrapper.trigger("keydown", { key: "Enter" });
  //     expect(wrapper.emitted("update:is-active")).toBeFalsy();

  //     const input = wrapper.find('input[type="checkbox"]');
  //     await input.trigger("change");
  //     expect(wrapper.emitted("update:is-active")).toBeFalsy();
  //   });

  //   test.each([true])(
  //     'should add class attribute "active" to container DOM if "state" prop equals %s',
  //     (state) => {
  //       const wrapper = mount(ToggleSwitch, {
  //         props: {
  //           state,
  //         },
  //       });

  //       const className = wrapper.classes(ACTIVE_CLASS);

  //       expect(className).toBe(true);
  //       expect(wrapper.classes(NORMAL_CLASS)).toBe(false);
  //       expect(className).toMatchSnapshot();
  //     }
  //   );

  //   test.each([false])(
  //     'should add class attribute "normal" to container DOM if "state" prop equals %s',
  //     (state) => {
  //       const wrapper = mount(ToggleSwitch, {
  //         props: {
  //           state,
  //         },
  //       });

  //       const className = wrapper.classes(NORMAL_CLASS);

  //       expect(className).toBe(true);
  //       expect(wrapper.classes(ACTIVE_CLASS)).toBe(false);
  //       expect(className).toMatchSnapshot();
  //     }
  //   );
  // });

  // describe("events", () => {
  //   test('should emit "update:is-acitve" when clicking on the wrapper', async () => {
  //     const wrapper = mount(ToggleSwitch, {
  //       props: {
  //         state: true,
  //         useNative: true,
  //       },
  //     });

  //     const inputEl = getInput(wrapper);
  //     await inputEl.trigger("change");

  //     const emittedEvent = getTypedEmittedEvent(wrapper, "update:is-active");

  //     expect(emittedEvent).toHaveLength(1);
  //     expect(wrapper.emitted()).toHaveProperty("update:is-active");
  //     expect(emittedEvent).toMatchSnapshot();
  //   });

  //   // test('should toggle state on "Enter" and "Space" keydown for custom switch', async () => {
  //   //   const wrapper = mount(ToggleSwitch, {
  //   //     props: { useNative: false, isActive: false },
  //   //   });

  //   //   await wrapper.trigger("keydown", { key: "Enter" });

  //   //   expect(wrapper.emitted("update:is-active")).toBeTruthy();

  //   //   await wrapper.trigger("keydown", { key: " " });

  //   //   expect(wrapper.emitted("update:is-active")?.length).toBe(2);
  //   // });
  // });
});
