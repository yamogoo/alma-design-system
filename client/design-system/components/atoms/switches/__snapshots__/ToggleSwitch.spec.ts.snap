// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleSwitch > events > should emit "update:is-acitve" when clicking on the wrapper 1`] = `
[
  [
    false,
  ],
]
`;
